src/ra_snapshot.erl

```diff
@@ -0,0 +1,354 @@
+-module(ra_snapshot).
+
+-include("ra.hrl").
+
+-type file_err() :: file:posix() | badarg | terminated | system_limit.
+-type meta() :: {ra_index(), ra_term(), ra_cluster_servers()}.
+
+-export([
+         save/4,
+         read/2,
+         recover/1,
+         read_meta/2,
+         delete/2,
+
+         init/3,
+         init_ets/0,
+         current/1,
+         pending/1,
+         accepting/1,
+         directory/1,
+         last_index_for/1,
+
+         begin_snapshot/3,
+         complete_snapshot/2,
+
+         begin_accept/4,
+         accept_chunk/3,
+         abort_accept/1,
+
+         handle_down/3
+        ]).
+
+-type effect() :: {monitor, process, snapshot_writer, pid()}.
+
+-export_type([meta/0, file_err/0, effect/0]).
+
+-record(accept, {num :: non_neg_integer(),
+                 %% the next expected chunk
+                 next = 1 :: non_neg_integer(),
+                 state :: term(),
+                 idxterm :: ra_idxterm()}).
+
+-record(?MODULE,
+        {uid :: ra_uid(),
+         module :: module(),
+         %% the snapshot directory
+         %% typically <data_dir>/snapshots
+         %% snapshot subdirs are store below
+         %% this as <data_dir>/snapshots/Term_Index
+         directory = file:filename(),
+         pending :: maybe({pid(), ra_idxterm()}),
+         accepting :: maybe(#accept{}),
+         current :: maybe(ra_idxterm())}).
+
+%% getter macro
+-define(GETTER(State), State#?MODULE.?FUNCTION_NAME).
+
+-define(ETSTBL, ra_log_snapshot_state).
+
+-opaque state() :: #?MODULE{}.
+
+-export_type([state/0]).
+
+%% Side effect function
+%% Turn the current state into immutable reference.
+-callback prepare(Index :: ra_index(),
+                  State :: term()) ->
+    Ref :: term().
+
+%% Saves snapshot from external state to disk.
+%% Runs in a separate process.
+%% External storage should be available to read
+-callback write(Location :: file:filename(),
+                Meta :: meta(),
+                Ref :: term()) ->
+    ok | {error, file_err() | term()}.
+
+%% Read the snapshot from disk into serialised structure for transfer.
+-callback read(ChunkSizeBytes :: non_neg_integer(),
+               Location :: file:filename()) ->
+    {ok, Meta :: meta(), ChunkState,
+     [ChunkThunk :: fun((ChunkState) -> {binary(), ChunkState})]} |
+    {error, term()}.
+
+%% begin a stateful snapshot acceptance process
+-callback begin_accept(SnapDir :: file:filename(),
```

Comment:
Can accept functions be in a separate module? I think most snapshotters just need to transfer a file of several files.

Reply:
no you need to be able to "connect both ends" the acceptor functions is the receiver of the chunks returned in read. how to interpret these chunks is snapshotter specific.

Reply:
I think there can be some helper functions to just write files.

Reply:
ok what can we add on top of `file:write_file/2` ?

Thinking about it should chunks be term() rather than binary() if you want to say, wrap it with some info around filenames and so forth?